bsp/src/org/jetbrains/bsp/project/importing/MillProjectInstaller.scala

```diff
@@ -16,25 +16,30 @@ class MillProjectInstaller extends BspProjectInstallProvider {
 
   override def canImport(workspace: File): Boolean =
     Option(workspace) match {
-      case Some(directory) if directory.isDirectory => isBspCompatible(directory) || isLegacyBspCompatible(directory)
+      case Some(directory) if directory.isDirectory =>
+        BspUtil.findFileByName(directory, "build.mill").isDefined ||
+          BspUtil.findFileByName(directory, "build.mill.scala").isDefined ||
+          isBspCompatible(directory) ||
+          isLegacyBspCompatible(directory)
       case _ => false
     }
 
   override def getConfigSetup: ConfigSetup = MillSetup
 
   override def serverName: String = "Mill"
 
-  override def installCommand(workspace: File): Try[Seq[String]] = {
+  override def installCommand(workspace: File): Try[Seq[String]] = Try {
     // note: The legacy part is only executed for mill bootstrap script so it is not applicable for Windows.
     // Maybe it could be, but we decided to support mill.bat file only for the newer bsp approach
     val isLegacyMill = !SystemInfo.isWindows && isLegacyBspCompatible(workspace)
     val millFileOpt = getMillFile(workspace)
     millFileOpt match {
-      case Some(file) if isMillFileBspCompatible(file, workspace) =>
-        Success(Seq(file.getAbsolutePath, "-i", "mill.bsp.BSP/install"))
-      case Some(file) if isLegacyMill =>
-        Success(Seq(file.getAbsolutePath, "-i", "mill.contrib.BSP/install"))
-      case _ => Failure(new IllegalStateException("Unable to install BSP as this is not a Mill project"))
+      case Some(file) if isLegacyMill && !isMillFileBspCompatible(file, workspace) =>
+        Seq(file.getAbsolutePath, "-i", "mill.contrib.BSP/install")
+      case Some(file) => // executes if isMillFileBspCompatible
```

Comment:
Now the comment:
```scala
// executes if isMillFileBspCompatible
```
is read as if the branch is only executed when `isMillFileBspCompatible == true`.
But AFAIU it's not the case: if `isLegacyMill` is `false` and `isMillFileBspCompatible` is `false` then the branch is still executed.

>is read as if the branch is only executed when `isMillFileBspCompatible == true`.

If it's indeed what is implied, it's better to return the original explicit check as it's less obfuscated.
If it's not what is implied, then the comment is probably not needed?

Reply:
I've improved the comment. It now gives a better idea what it's aiming for.

The code itself is now more robust. It tries as much as possible to import the Mill build, with the best the user has.